0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,694 @@
+  /// the nodes in the new committee.
+  ///
+  /// Full share:
+  ///      s'(y) = B'(node_id,y)
+  fn full_share_distribution(&self, churp_id: u8, round: u64);
+
+  /// Return switch data point (integer) for the calling node.
+  /// 
+  /// The point is evaluation of the shared secret (polynomial)
+  /// at the given y value (caller id).
+  /// 
+  /// Switch point:
+  ///     P = g(caller_id) = B(node_id,caller_id)
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the polynomial needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn switch_point(&self, churp_id: u8, round: u64) -> Integer;
+
+  /// Return the key share for the given round.
+  ///
+  /// To construct a derived key, the caller is required to collect 
+  /// at least the threshold number of key shares and combine them
+  /// locally to construct the derived key.
+  ///
+  /// Key share:
+  ///     K = H(runtime_id, key_pair_id)^{r g(0)}
+  ///
+  ///
+  /// The caller should always provide a proof which can be independently
+  /// verified before key shares are released to the caller.
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the key share needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn key_share(&self, churp_id: u8, round: u64, runtime_id: Namespace, 
+    key_pair_id: KeyPairId, proof: Proof) -> Integer;
+}
+```
+
+Methods:
+
+- `Churp.Init`
+
+- `Churp.VerificationMatrix`
+
+- `Churp.BivariateShare`
+
+- `Churp.Dealing`
+
+- `Churp.ShareReduction`
+
+- `Churp.ProactiveRandomization`
+
+- `Churp.FullShareDistribution`
+
+- `Churp.SwitchPoint`
+
+- `Churp.KeyShare`
+
+### Key manager client
+
+```rust
+/// Key manager client for CHURP.
+pub trait Client {
+  /// Return the key for the given round.
+  ///
+  /// The key is combined from key shares obtained from the key manager nodes
+  /// in the current CHURP committee.
+  fn key(&self, churp_id: u8, round: u64, runtime_id: Namespace, key_pair_id: KeyPairId) -> Vec<u8>;
+}
+```
+
+## Key Derivation Center
+
+Key Derivation Center (KDC) is a secret sharing scheme based on
+the verifiable secret sharing scheme (VSS) where every node possesses
+only a share of the master secret. To derive a key from the master secret,
+one needs to obtain at least t+1 threshold number of key shares from distinct
+nodes and reassemble them locally.
+
+```rust
+/// Key manager application that implements key derivation center (KDC).
+pub trait KDC {
+  /// TODO: Define methods if we decide to implement KDC also.
+}
+```
+
+## Consequences
+
+### Positive
+
+CHURP:
+
+- High security, as the master secret is shared among key manager nodes.
+
+- Supports proactive randomization (share refresh).
+
+- Dynamic committees.
+
+KDC:
+
+- High security, as the master secret is shared among key manager nodes.
+
+- Supports proactive randomization (share refresh).
+
+### Negative
+
+CHURP:
+
+- Handoffs are computationally intensive.
```

Comment:
How intensive (in terms of latency/communication rounds)?

Reply:
Didn't want to be exact, as handoff is more complex. Roughly, it needs 2 x consensus transactions, 2 x communication rounds, 2 x Lagrange interpolations, N x SHA computations, and a bunch of group operations (multiplication or exponentiation).